Restart from failure

dpguthrie opened the pull request.

matt-winkler commented on dbtc/client/cloud/base.py:
```diff
@@ -968,46 +975,172 @@ def test_connection(self, account_id: int, payload: Dict) -> Dict:
+
+                        # errors and failures are when we need to inspect to figure
+                        # out the point of failure
+                        elif step_meets_condition(run_step, 'run', ['error', 'failed']):
```
@dpguthrie this is looking like a solid step forward! One other thing occurred to me: `--vars`. For non-run (e.g. anything other than `dbt build` and `dbt run`) and skipped steps, it looks like this is copying the `--vars` statements correctly. But on failed steps like `dbt test` or the point of failure step, those extra arguments are being dropped.

Noting also that I don't think we need to worry about replicating any `--selectors`, `--exclude` etc. because those statements will scope the initial run, and we're just working within that scope for the restart from failure.

matt-winkler commented:
Hey @dpguthrie I did some work locally to make this handle appending `--args` and `--vars` to the restart command. But, I'm blocked from making commits to your repo.

dpguthrie commented:
@matt-winkler This is awesome!  I saw the `vars` piece in the googlesheet you shared but wasn't sure how we'd get that - but this looks like it works!  I agree with not needing to pull in `select` or `exclude` as we're already getting that info with the `rerun_nodes`, but what about `defer` and `state`?

matt-winkler commented:
Yeah I think our rule is "anything that deals with node selection / scoping on the previous run DOES NOT need to be repeated by the logic in the restart"

dpguthrie commented:
@matt-winkler I added your code above mostly verbatim except the following:

- Changed where the argument parser is initialized (only when we get into restart from failure)
- Added single quotes around parsed_command.args and vars

Then added a couple tests that ensure that we're getting the expected override steps (both are passing).
